File: orison-web/src/engine/scene.js
export class Scene {
  constructor({ name = 'Scene' } = {}) {
    this.name = name;
    this.game = null;

    this._entities = [];
    this._toAdd = [];
    this._toRemove = new Set();

    this.paused = false;

    this.cameraX = 0;
    this.cameraY = 0;
    this.cameraZoom = 1;
  }

  begin(game) {
    this.game = game;
  }

  end() {
    for (const e of this._entities) e.removed();
    this._entities = [];
    this._toAdd = [];
    this._toRemove.clear();
    this.game = null;
  }

  add(entity) {
    this._toAdd.push(entity);
    return entity;
  }

  remove(entity) {
    this._toRemove.add(entity);
  }

  _flushQueues() {
    if (this._toAdd.length) {
      for (const e of this._toAdd) {
        e.added(this);
        this._entities.push(e);
      }
      this._toAdd.length = 0;
    }

    if (this._toRemove.size) {
      const removeSet = this._toRemove;
      this._entities = this._entities.filter((e) => {
        if (!removeSet.has(e)) return true;
        e.removed();
        return false;
      });
      removeSet.clear();
    }
  }

  update(dt) {
    this._flushQueues();
    if (this.paused) return;

    for (const e of this._entities) {
      if (!e.active) continue;
      e.update(dt);
    }
  }

  render(renderer) {
    const g = renderer;
    g.pushCamera(this.cameraX, this.cameraY, this.cameraZoom);

    for (const e of this._entities) {
      if (!e.visible) continue;
      e.render(g);
    }

    g.popCamera();
  }
}

File: src/Orison.Core/Core/Scene.cs
using System;
using System.Collections.Generic;

namespace Orison {
    /// <summary>
    /// A Scene is a container for Entities. Scenes are managed by the Game class.
    /// </summary>
    public class Scene {
        #region Public Fields

        /// <summary>
        /// The list of Entities in the Scene.
        /// </summary>
        public List<Entity> Entities {
            get;
            private set;
        }

        #endregion

        #region Public Properties

        /// <summary>
        /// The Game the Scene is currently in.
        /// </summary>
        public Game Game {
            get;
            internal set;
        }

        #endregion

        #region Constructors

        /// <summary>
        /// Create a new Scene.
        /// </summary>
        public Scene() {
            Entities = new List<Entity>();
        }

        #endregion

        #region Public Methods

        /// <summary>
        /// Add an Entity to the Scene.
        /// </summary>
        /// <param name="entity">The Entity to add.</param>
        public T Add<T>(T entity) where T : Entity {
            entity.Scene = this;
            Entities.Add(entity);
            entity.Added();
            return entity;
        }

        /// <summary>
        /// Remove an Entity from the Scene.
        /// </summary>
        /// <param name="entity">The Entity to remove.</param>
        public void Remove(Entity entity) {
            if (Entities.Contains(entity)) {
                entity.Removed();
                Entities.Remove(entity);
            }
        }

        /// <summary>
        /// Get the first Entity with a specific name.
        /// </summary>
        /// <param name="name">The name to search for.</param>
        /// <returns>The Entity if found, null otherwise.</returns>
        public Entity GetEntity(string name) {
            foreach (var entity in Entities) {
                // Note: Entity doesn't have a Name property in the original,
                // but we could add it or use type checking
                if (entity.GetType().Name == name) return entity;
            }
            return null;
        }

        /// <summary>
        /// Get all Entities of a specific type.
        /// </summary>
        /// <typeparam name="T">The type of Entity to get.</typeparam>
        /// <returns>A list of Entities of the specified type.</returns>
        public List<T> GetEntities<T>() where T : Entity {
            var list = new List<T>();
            foreach (var entity in Entities) {
                if (entity is T) list.Add((T)entity);
            }
            return list;
        }

        #endregion

        #region Virtual Methods

        /// <summary>
        /// Called when the Scene is added to the Game.
        /// </summary>
        public virtual void Begin() {
        }

        /// <summary>
        /// Called when the Scene is removed from the Game.
        /// </summary>
        public virtual void End() {
            foreach (var entity in Entities) {
                entity.Removed();
            }
            Entities.Clear();
        }

        /// <summary>
        /// Called every frame to update the Scene.
        /// </summary>
        public virtual void Update() {
            foreach (var entity in Entities) {
                if (entity.AutoUpdate) {
                    entity.UpdateFirst();
                    entity.Update();
                    entity.UpdateLast();
                }
            }
        }

        /// <summary>
        /// Called every frame to render the Scene.
        /// </summary>
        public virtual void Render() {
            foreach (var entity in Entities) {
                if (entity.AutoRender && entity.Visible) {
                    entity.Render();
                }
            }
        }

        #endregion
    }
}
